src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,339 @@
+                _examinedEverything = examinedIndex == _readTail.End - _readTail.Start;
+            }
+
+            // Three cases here:
+            // 1. All data is consumed. If so, we clear _readHead/_readTail and _readIndex/
+            //  returnEnd is set to null to free all memory between returnStart/End
+            // 2. A segment is entirely consumed but there is still more data in nextSegments
+            //  We are allowed to remove an extra segment. by setting returnEnd to be the next block.
+            // 3. We are in the middle of a segment.
+            //  Move _readHead and _readIndex to consumedSegment and index
+            if (_bufferedBytes == 0)
+            {
+                _readTail.SetMemory(_readTail.MemoryOwner);
+                _readHead = _readTail;
+                returnEnd = _readTail;
+                _readIndex = 0;
+            }
+            else if (consumedIndex == returnEnd.Length)
+            {
+                var nextBlock = returnEnd.NextSegment;
+                _readHead = nextBlock;
+                _readIndex = 0;
+                returnEnd = nextBlock;
+            }
+            else
+            {
+                _readHead = consumedSegment;
+                _readIndex = consumedIndex;
+            }
+
+            // Remove all blocks that are freed (except the last one)
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnStart.ResetMemory();
+                returnStart = returnStart.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException("OnWriterCompleted is not supported");
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
```

Comment:
It's possible we'll want to move this code path out of the async method to avoid the state machine if we can. (perf for later)